In [30]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler

In [31]:
#cargamos el dataset
df=pd.read_csv('datos_sin_diabetes.csv')
df.head(10)

#Los q pienso q pueeden tener diabetes son: 8,4,3,6


Embarazos  Glucosa  Presion_Arterial  Espesor_Piel  Test Insulina   IMC  \
0          6      148                72            35              0  33.6   
1          1       85                66            29              0  26.6   
2          8      183                64             0              0  23.3   
3          1       89                66            23             94  28.1   
4          0      137                40            35            168  43.1   
5          5      116                74             0              0  25.6   
6          3       78                50            32             88  31.0   
7         10      115                 0             0              0  35.3   
8          2      197                70            45            543  30.5   
9          8      125                96             0              0   0.0   

   DiabetesPedigree  Edad  
0             0.627    50  
1             0.351    31  
2             0.672    32  
3             0.167    21  
4             2.288    33  
5             0.201    30  
6             0.248    26  
7             0.134    29  
8             0.158    53  
9             0.232    54

# Actividad 6
## Parte 2: Implementación en Python
### Paso 1: Carga y Exploración de los Datos (20 min)
Los alumnos deben:
- Cargar los datos en un DataFrame de pandas.
- Inspeccionar las primeras filas.
- Obtener estadísticas descriptivas (media, desviación estándar, valores faltantes).
Código base:


In [36]:
columnas = ["Embarazos", "Glucosa", "Presion_Arterial", "Espesor_Piel", "Insulina", "IMC", "DiabetesPedigree", "Edad", "Diabetes"]
data=pd.read_csv('pima-indians-diabetes.data.csv', names=columnas)
print('Contenido del dataset:\n')
data.head(10)


Contenido del dataset:



Embarazos  Glucosa  Presion_Arterial  Espesor_Piel  Insulina   IMC  \
0          6      148                72            35         0  33.6   
1          1       85                66            29         0  26.6   
2          8      183                64             0         0  23.3   
3          1       89                66            23        94  28.1   
4          0      137                40            35       168  43.1   
5          5      116                74             0         0  25.6   
6          3       78                50            32        88  31.0   
7         10      115                 0             0         0  35.3   
8          2      197                70            45       543  30.5   
9          8      125                96             0         0   0.0   

   DiabetesPedigree  Edad  Diabetes  
0             0.627    50         1  
1             0.351    31         0  
2             0.672    32         1  
3             0.167    21         0  
4             2.288    33         1  
5             0.201    30         0  
6             0.248    26         1  
7             0.134    29         0  
8             0.158    53         1  
9             0.232    54         1

In [37]:
print('Estadística\n', data.describe())
print('Forma del dataset\n', data.shape)
data.isnull().sum() 
#print(data.info())
#No parecen haber valores nulos.
#Hay algunos valores extremos, como por ejemplo en la columna de insulina, donde le max es 846
#Se podrían elimnar estos valores extremos.

Estadística
         Embarazos     Glucosa  Presion_Arterial  Espesor_Piel    Insulina  \
count  768.000000  768.000000        768.000000    768.000000  768.000000   
mean     3.845052  120.894531         69.105469     20.536458   79.799479   
std      3.369578   31.972618         19.355807     15.952218  115.244002   
min      0.000000    0.000000          0.000000      0.000000    0.000000   
25%      1.000000   99.000000         62.000000      0.000000    0.000000   
50%      3.000000  117.000000         72.000000     23.000000   30.500000   
75%      6.000000  140.250000         80.000000     32.000000  127.250000   
max     17.000000  199.000000        122.000000     99.000000  846.000000   

              IMC  DiabetesPedigree        Edad    Diabetes  
count  768.000000        768.000000  768.000000  768.000000  
mean    31.992578          0.471876   33.240885    0.348958  
std      7.884160          0.331329   11.760232    0.476951  
min      0.000000          0.078000   21.0000

Embarazos           0
Glucosa             0
Presion_Arterial    0
Espesor_Piel        0
Insulina            0
IMC                 0
DiabetesPedigree    0
Edad                0
Diabetes            0
dtype: int64

### Paso 2: Preparación de Datos
- Separar características (X) y variable objetivo (y).
-  Normalizar los datos con StandardScaler.
- Dividir en entrenamiento y prueba (80%-20%).
Código:


In [ ]:
# Separar características y variable objetivo

X = data.drop(columns=["Diabetes"]) #Características
y = data["Diabetes"] #Variable objetivo

# Dividir en entrenamiento y prueba (20% para prieba)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# Normalización
# Se normalizan los datos para que tengan una media de 0 y una desviación estándar de 1.
scaler = StandardScaler() 
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
# Entrenar modelo Random Forest

modelo = RandomForestClassifier(n_estimators=100, random_state=42)
modelo.fit(X_train, y_train)


# Predicciones

y_pred = modelo.predict(X_test)
# Imprimir las predicciones
print("Predicciones del modelo:")
print(y_pred)


# Comparar predicciones con los valores reales
resultados_comparacion = pd.DataFrame({'Real': y_test, 'Predicción': y_pred})
print(resultados_comparacion.head(10))  # Muestra las primeras 10 comparaciones


# Evaluación
accuracy = accuracy_score(y_test, y_pred)
matriz_confusion = confusion_matrix(y_test, y_pred)
print(f"Precisión del modelo: {accuracy:.2f}")
print(f"Matriz de Confusión:{matriz_confusion}")




Predicciones del modelo:
[0 0 0 0 0 1 0 1 1 1 0 1 0 0 0 0 0 0 1 1 0 0 0 0 1 1 0 0 0 0 1 1 1 1 1 1 1
 0 0 1 0 0 1 0 0 1 0 0 0 1 0 1 1 0 0 0 1 0 0 1 1 0 0 0 0 1 0 0 0 1 1 0 0 0
 0 1 0 0 0 0 1 0 0 1 0 1 1 0 0 0 0 0 0 1 0 1 1 0 1 0 1 0 0 1 1 0 0 1 0 1 0
 1 0 1 0 1 1 0 0 1 0 0 0 0 0 0 0 1 1 1 1 1 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0
 0 1 0 0 0 0]
     Real  Predicción
668     0           0
324     0           0
624     0           0
690     0           0
473     0           0
204     0           1
97      0           0
336     0           1
568     0           1
148     0           1
Precisión del modelo: 0.72
Matriz de Confusión:[[77 22]
 [21 34]]


Preguntas para los alumnos:
1. ¿El modelo es suficientemente preciso para ser usado en un entorno médico?
La precisión es muy baja (0.72)

2. ¿Qué pasaría si hubiera más datos o datos de mejor calidad?


3. ¿Cómo podríamos mejorar el modelo? (Ejemplo: probar otro algoritmo, ajustar hiperparámetros)